core/shared/src/main/scala/scalaz/zio/duration/Duration.scala

```diff
@@ -0,0 +1,73 @@
+package scalaz.zio.duration
+
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.{ Duration => ScalaDuration, FiniteDuration => ScalaFiniteDuration }
+import scala.math.Ordered
+
+sealed trait Duration extends Ordered[Duration] {
+
+  def +(other: Duration): Duration
+
+  def *(factor: Double): Duration
+
+  def max(other: Duration): Duration = if (this > other) this else other
+
+  def min(other: Duration): Duration = if (this < other) this else other
+
+}
+
+final object Duration {
+
+  object Finite {
+
+    def apply(nanos: Long): Duration =
+      if (nanos >= 0) new Finite(nanos)
+      else Infinity
+
+  }
+
+  final case class Finite private (nanos: Long) extends Duration {
+
+    def +(other: Duration): Duration = other match {
+      case Finite(otherNanos) => Finite(nanos + otherNanos)
+      case Infinity           => Infinity
+    }
+
+    def *(factor: Double): Duration =
+      if (!factor.isInfinite && !factor.isNaN) Finite((nanos * factor).round)
+      else Infinity
+
+    def compare(other: Duration) = other match {
+      case Finite(otherNanos) => nanos compare otherNanos
+      case Infinity           => -1
+    }
+
+    def copy(nanos: Long = this.nanos): Duration = Finite(nanos)
```

Comment:
This allows the user to introduce negative nanos.

We may need to not use `case class` / `case object` and rather define our own `hashCode` and `equals`.

Reply:
Actually this hides the `copy` method from the case class and delegates to `Finite.apply` from the companion object (which in turn takes care of the < 0 case). So it's safe.